lisp/ess-s-lang.el

```diff
@@ -575,68 +575,44 @@ and one that is well formatted in emacs ess-mode."
     (ess-if-verbose-write "ess-fix-misc __end__\n");___D___
     ))
 
-(defvar polymode-mode)
-(defun ess-smart-S-assign (&optional N)
-  "Act as smart `ess-S-assign' key.
-Insert `ess-S-assign', unless in string/comment.  If the
-underscore key is pressed a second time, the assignment operator
-is removed and replaced by the underscore.  `ess-S-assign',
-typically \" <- \", can be customized.  In ESS modes other than
-R/S, the underscore is always inserted.
-
-If `ess-smart-S-assign-key' is nil, just call
-`self-insert-command'. You can pass N as usual."
+(defun ess-cycle-assignment ()
+  "Cycle between assignment symbols in `ess-assign-list'.
+On consecutive calls, replace the assignment symbol before point
+with the next symbol from that list."
   (interactive)
-  (if ess-smart-S-assign-key
-      (save-restriction
-        (ignore-errors
-          (when (and (eq major-mode 'inferior-ess-mode)
-                     (> (point) (process-mark (get-buffer-process (current-buffer)))))
-            (narrow-to-region (process-mark (ess-get-process)) (point-max)))
-          (and ess-noweb-mode
-               (ess-noweb-in-code-chunk)
-               (ess-noweb-narrow-to-chunk))
-          (and (fboundp 'pm/narrow-to-span)
-               polymode-mode
-               (pm/narrow-to-span)))
-        (if (or
-             (ess-inside-string-or-comment-p (point))
-             (not (equal ess-language "S")))
-            (insert ess-smart-S-assign-key)
-          (ess-insert-S-assign)))
-    (funcall #'self-insert-command (or N 1))))
+  (if (eq last-command this-command)
+      (let ((slist ess-assign-list)
+            str)
+        (while (and (setq str (car slist)
+                          slist (cdr slist))
+                    (not (and (re-search-backward str
+                                                  (- (point) (length str)) t)
```

Comment:
BTW, I think whitespace cleanup around point might be useful here.

Reply:
I was thinking that too, but am not sure how to handle all situations. Some people like to align on the assignment operator, which means that a lot of whitespace might be desirable:

```
this    <- "that"
that    <- "this"
longer  <- "this_or_that"
```

Reply:
Good point. Let's leave it as it is then.